### PR TITLE
feat(crypto): add KeyStack for rotatable key data signing

### PR DESCRIPTION
**Pull request by kitsonk**
This adds a utility class which is a pre-cursor to being able to provide a cookie interface that can transparently sign and validate client cookies to prevent cookie tampering, which is a best practice we should be promoting.  It is currently part of `oak_commons` that I am contributing back to `std`.

### CODE REVIEW

File: crypto/keystack.ts [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+  /** Sign the data, returning a string based digest of the data. */
+  sign(data: Data): Promise<string> | string;
+  /** Verifies the digest matches the provided data, indicating the data was
+   * signed by the keyring and has not been tampered with. */
+  verify(data: Data, digest: string): Promise<boolean> | boolean;
+}
+
+const encoder = new TextEncoder();
+
+function assert(value: unknown, message = "Assertion error"): asserts value {
```
Comment (kt3k) [inline crypto/keystack.ts:35]: How about using one in `std/_util/assert.ts`?

File: crypto/keystack.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+/** Types of data that can be signed cryptographically. */
+export type Data = string | number[] | ArrayBuffer | Uint8Array;
+
+/** Types of keys that can be used to sign data. */
+export type Key = string | number[] | ArrayBuffer | Uint8Array;
+
+/** An abstract interface for a keyring which handles signing of data based on
+ * a string based digest. */
+export interface KeyRing {
```
Comment (kt3k) [inline crypto/keystack.ts:21]: Why does this have a different name from the class name? How about naming both KeyStack?

File: crypto/keystack.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  }
+}
+
+const REPLACEMENTS: Record<string, string> = {
+  "/": "_",
+  "+": "-",
+  "=": "",
+};
+
+function encodeBase64Safe(data: string | ArrayBuffer): string {
```
[the remark this answers is not in the document]
Comment (kitsonk) [reply crypto/keystack.ts:47]: 👍  will do

File: crypto/keystack.ts (hunk 1 above)
Comment (kitsonk) [reply crypto/keystack.ts:35]: sorry, yeah, should have refactored that.

File: crypto/keystack.ts (hunk 2 above)
Comment (kitsonk) [reply crypto/keystack.ts:21]: As stated, it is an abstract interface of which `KeyStack` is _an_ implementation of it. What we are establishing here though is a generic interface people would then be able to produce their own implementations if they wanted to.  There isn't _yet_ a good home for this interface outside of here at the moment.

Comment (kt3k) [reply crypto/keystack.ts:21]: Ok. Sounds fair enough

**Review by kt3k — APPROVED** [review]
LGTM 👍

**Pull request merged**